mantidproject/mantid
# Specify input workspace flipper configuration order

Fixes #36146.
First part of #35067. When calling `PolarizationCorrectionWildes`, one can now specify any order of the input workspaces, they do not have to be in one of the hard-coded orders in the algorithm.

#### Summary of work

The `PolarizationCorrectionWildes` algorithm takes in various numbers of workspaces, depending on the analysis being conducted and whether or not an analyser is present. The flipper configuration is specified by a string, validated using the existing `SpinStateValidator` class. The main change I've made is where the input workspaces are processed and assigned to the four workspaces on the input `WorkspaceMap` object.

I've added the ability to `SpinStateValidator` to take single spin states, e.g. `0,1`, which is what you get if no analyser is present.

To clarify the difference between spin state and flipper configuration, a neutron can either be up or down spin. A flipper will change the spin of the neutron from one spin state to the other. Hence in this case the output workspaces appear with `+,-` in their names, signifying the neutron spin. A `0,1` tells you if the flipper is on or off, so to go from that to a neutron spin you need to know the spin of the incoming neutron beam, in this case assumed up (`+`) I think. I didn't change any of that logic though, so this is just for info.

### To test:
Should be completely backwards compatible so all old tests and scripts should work. I've added some tests that reorder the input workspaces, change the corresponding `Flippers` input string, and check that all the results are still the same.







---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
